### PR TITLE
Code Review:  DPPTSBGTC-2673

### PR DESCRIPTION
**Pull request by kevinjqiu**
Just for code review, not meant for merge.

https://jira.points.com/browse/DPPTSBGTC-2673

[Review on Reviewable](https://reviewable.io/reviews/points/recheck/1)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+language: python
```
Comment (gdmello) [inline .travis.yml]: travis for CI?

File: recheck/cli.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    requirements_parser = requirements.RequirementsParser(requirements_file)
+
+    ignored_requirements = requirements.get_ignored_requirements(ignore_file)
+
+    ignored, outdated_major, outdated_minor = set(), set(), set()
+    for line in _list_oudated_requirements(requirements_parser.index_url,
+                                           requirements_parser.extra_index_urls):
+        ui.progress()
+        req = requirements.parse_result(line)
```
Comment (gdmello) [inline recheck/cli.py]: avoid abbreviations.

File: recheck/cli.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        if not req:
+            # the output does not resemble an outdated requirement
+            continue
+
+        if req.name not in requirements_parser.direct_requirements:
+            # not a direct requirement
```
Comment (gdmello) [inline recheck/cli.py]: the conditional suffices in place of the comment :)

Comment (kevinjqiu) [inline recheck/cli.py]: Done https://github.com/Points/recheck/commit/12937fdd70608c62ed146b22ea0973a0b0e978b1

File: recheck/cli.py (hunk 2 above)
Comment (kevinjqiu) [inline recheck/cli.py]: Done https://github.com/Points/recheck/commit/12937fdd70608c62ed146b22ea0973a0b0e978b1

File: .travis.yml (hunk 1 above)
Comment (kevinjqiu) [inline .travis.yml]: Yeah it's free for public repos.

Comment (gdmello) [inline .travis.yml]: Should we change this to TC for points?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	git push --tags
+
+build:
+	python setup.py sdist bdist_wheel
+
+upload:
+	python setup.py bdist_wheel upload
+
+testbuild:
+	py.test tests && make build
```
Comment (biniambekele) [inline Makefile]: Could also use ``make test``

Comment (kevinjqiu) [inline Makefile]: Will remove this target.  It's temporary.

File: recheck/cli.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
+import subprocess
+import sys
+
+import click
+from recheck import requirements, textui as ui
```
Comment (biniambekele) [inline recheck/cli.py]: I'd suggest either renaming "textui" to "ui" or importing it as "textui" (w/o the alias).

File: recheck/requirements.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    def __eq__(self, other):
+        return all([
+            self.name == other.name,
+            self.installed_version == other.installed_version,
+            self.remote_version == other.remote_version,
+            self.requirements_file == other.requirements_file,
+        ])
+
+    def __hash__(self):
+        return reduce(operator.xor, map(hash, [self.name,
```
Comment (biniambekele) [inline recheck/requirements.py]: Alternatively: 
```python
return hash((self.name, self.installed_version, self.remote_version, self.requirements_file))
```

File: recheck/requirements.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                                               self.installed_version,
+                                               self.remote_version,
+                                               self.requirements_file]))
+
+
+Version = collections.namedtuple('Version', ['major', 'minor', 'rev'])
```
Comment (biniambekele) [inline recheck/requirements.py]: I'd expect "rev" to be named "patch" [as per semver](http://semver.org/).

File: recheck/requirements.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        if directive == '-r':
+            filepath = value.strip()
+            if not os.path.isabs(filepath):
+                filepath = os.path.join(self._dirname, filepath)
+            self._requirements_files.append(filepath)
+        if directive == '--index-url':
+            self._index_url = value.strip()
+        if directive == '--extra-index-url':
+            self._extra_index_urls.append(value.strip())
+        if directive == '-e':
```
Comment (biniambekele) [inline recheck/requirements.py]: Suggest: {{elif}}

File: recheck/requirements.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        name, info = line.strip().split(' ', 1)
+        _, installed_version, _, remote_version = info[1:-1].split(' ')[:4]
+        return OutdatedRequirement(name, installed_version, remote_version)
+    except ValueError:
+        return None
+
+
+def get_ignored_requirements(ignore_file):
+    if not os.path.exists(ignore_file):
+        return set([])
```
Comment (biniambekele) [inline recheck/requirements.py]: ``set()`` also works.

File: setup.py [hunk 10]
```diff
@@ -1,26 +1,33 @@
...
-    version="0.1.0",
+    version=version,
     url="https://github.com/kevinjqiu/recheck",
 
     author="Kevin J. Qiu",
-    author_email="kevin@idempotent.ca",
+    author_email="kevin.qiu@points.com",
 
     description="Re(quirements)Check",
     long_description=open('README.rst').read(),
```
Comment (biniambekele) [inline setup.py]: Could extract this to be consistent with "install requires" and "version". Could also ``strip()`` all three.

**Comment by kevinjqiu** [comment]
output

**Comment by biniambekele** [comment]
Very nice. LGTM.

File: recheck/requirements.py (hunk 9 above)
Comment (kevinjqiu) [inline recheck/requirements.py]: Done. https://github.com/Points/recheck/commit/3a8119f927247f8138ce8e6795fc5181f620abbc

File: setup.py (hunk 10 above)
Comment (kevinjqiu) [inline setup.py]: Done. https://github.com/Points/recheck/commit/3a8119f927247f8138ce8e6795fc5181f620abbc

File: recheck/requirements.py (hunk 8 above)
Comment (kevinjqiu) [inline recheck/requirements.py]: Done. https://github.com/Points/recheck/commit/3a8119f927247f8138ce8e6795fc5181f620abbc

File: recheck/requirements.py (hunk 7 above)
Comment (kevinjqiu) [inline recheck/requirements.py]: Done. https://github.com/Points/recheck/commit/3a8119f927247f8138ce8e6795fc5181f620abbc

File: recheck/cli.py (hunk 5 above)
Comment (kevinjqiu) [inline recheck/cli.py]: Done. https://github.com/Points/recheck/commit/3a8119f927247f8138ce8e6795fc5181f620abbc

File: Makefile (hunk 4 above)
Comment (kevinjqiu) [inline Makefile]: Done. https://github.com/Points/recheck/commit/3a8119f927247f8138ce8e6795fc5181f620abbc

File: .travis.yml (hunk 1 above)
Comment (kevinjqiu) [inline .travis.yml]: Sure, although this file doesn't affect TC build.

**Pull request closed without merging**